### ISSUE TITLE
issue посвящённый проблемам 31-го февраля (и 32-го мая...)

### ISSUE THREAD

**Issue by uliss4ever** [uliss4ever/My_repository_2#5]
>      def __init__(self, *args):

self._year=нет, self._month=нет, self._day=нет

Первая строчка вообще ничего не меняет:
>          """ ВАЖНО! ..."""

self._year=нет, self._month=нет, self._day=нет
@Lex-2008
Лёш, с какого момента _year, _month и _day материализуются? Я исключительно о нижнем подчеркивании. Это то, что нам пришло в голову в роли args ? Любые числа? Вроде я всё про приватный аргумент помню, но откуда именно он берётся, просто из нашей головы, да?
И вот я дошла, месяц в сеттере он устанавливает. Строчкой выше - у меня год. Подсказок нет)
self.year = int(args[2])
self.month = int(args[1])
self.day = int(args[0])
И ещё скажи, в self.day и day() из сеттера это не одно и то же? Одно атрибут, другое функция. Но как-то же они соотносятся. Их можно по-разному назвать? (мой ответ - да, вроде)

**Comment by uliss4ever** [comment]
Да, с трависом бодалась. Вроде победила, но не знаю, почему код 1, а не 0. Тесты упали, да, но они же отработали? 
![Снимок экрана 2021-02-28 в 17 23 10](https://user-images.githubusercontent.com/32711430/109423051-1f0ca880-79ef-11eb-8b8a-da5dbb6513b2.png)

**Comment by Lex-2008** [comment]
Да, всё правильно: Тесты отработали и упали - потому что в коде ошибка. Для этого тесты и нужны - чтобы падать, когда ошибка. Ура!

Сейчас ещё напишу.

**Comment by Lex-2008** [comment]
> с какого момента _year, _month и _day материализуются?

Вот, это правильный вопрос! Ну а ответ на него простой: тогда, когда они стоят **слева** от знака присваивания (`=`). Если поискать по коду, то находится только по одному месту для каждого:

https://github.com/uliss4ever/My_repository_2/blob/7c1c7a4c237b2a98d013c875e3b0c3c2ab341e38/class_Date.py#L87-L92

https://github.com/uliss4ever/My_repository_2/blob/7c1c7a4c237b2a98d013c875e3b0c3c2ab341e38/class_Date.py#L98-L103

https://github.com/uliss4ever/My_repository_2/blob/7c1c7a4c237b2a98d013c875e3b0c3c2ab341e38/class_Date.py#L114-L118

И в каждом случае перед этим вызывается (или _должна_ вызываться) функция `is_valid_date`.

Ты его вызываешь правильно, и передаёшь в него правильные агрументы.

> self._year=нет, self._month=нет, self._day=нет
> И вот я дошла, месяц в сеттере он устанавливает. Строчкой выше - у меня год. Подсказок нет)
> self.year = int(args[2])
> self.month = int(args[1])

Правильно, давай сначала год разберем. Это не та подсказка, о которой я говорил - подсказкой будет вот эта строчка:

> self._year=нет, self._month=нет, self._day=нет

которая говорит, что _пока_ у нас нет ни self._year, ни self._month, ни self._day.

Итак, у нас нет ни self._year, ни self._month, ни self._day, и мы (ну, то есть, интерпретатор Питона) собирается выполнить вот эту строчку:

> self.year = int(args[2])

Что в ней записано? Что надо вычислить "int(args[2])", и записать результат в "self.year". "self" - это у нас объект класса Date (мы сейчас init разбираем, помнишь?), и year - это его свойство. Свойство - значит, когда мы пытаемся туда что-то записать, интерпретатор вызывает сеттер. Сеттер у нас начинается такими словами:
````
    @year.setter 
    def year(self, value: int): 
````
Это значит, что setter - это просто функция, которой передается два параметра: "self" - тот объект, свойство которого мы хотим изменить, и "value" - то значение, которое мы хотим туда записать. Снова вспоминаем, что мы разбираем функцию init, а конкретно вот эту строчку:
> self.year = int(args[2])

Значит, self в сеттере - это тот же self, что в этой строчке, а value - то, что сосчитает нам "int(args[2])" (ну, какое-то число).

Отлично, и что же этот сеттер делает?

Он должен делать вот это:
````
        if not self.is_valid_date(self.day, self.month, value):
            raise ValueError
        self._year = value
````

Расскажешь, что он делает в первой строчке?

**Comment by Lex-2008** [comment]
> Это то, что нам пришло в голову в роли args ? Любые числа? Вроде я всё про приватный аргумент помню, но откуда именно он берётся, просто из нашей головы, да?

Да, если мы посмотрим на строчку, как этот init вызывается:
https://github.com/uliss4ever/My_repository_2/blob/7c1c7a4c237b2a98d013c875e3b0c3c2ab341e38/class_Date.py#L29

то там написано, что args - это все аргументы, которые передаются в init, кроме self. Ну, то есть, просто все агрументы, которые мы задаём, когда пишем `d1 = Date(…)`.

> И ещё скажи, в self.day и day() из сеттера это не одно и то же? Одно атрибут, другое функция. Но как-то же они соотносятся. Их можно по-разному назвать? (мой ответ - да, вроде).

Да, можно по-разному. Вот [здесь][d] есть другой пример, как property устанавливать: сначала объявляем любые три функции, которые читают/записывают/удаляют какую-то внутреннюю переменную, а потом создаем свойство, которое будет их вызывать. `@property` - это просто другой способ записи того же, более "красивый" (утверждение спорное, т.к. не всем он нравится).

[d]: https://docs.python.org/3/library/functions.html#property

**Comment by uliss4ever** [comment]
> Расскажешь, что он делает в первой строчке?

Он идёт и проверяет, то, что мы ему подали (self.day, self.month, value) - будет ли нормальной датой?
Очень-очень понятно! Только можно я попробую продолжить и объяснить, почему если мы сделали такую проверку, у нас инит свалился?
У нас есть дата 31/01/2021. Она задана в ините (да?) и 31 - это _day, 01 - _month, 2021 -_year, только устанавливали в обратном порядке.
И мы ему сказали, проверь, пожалуйста, 31/02/2021 будет ли валидной датой? Для этого нужен опять сеттер. Который идет и вызывает вот это def is_valid_date(cls, day: int, month: int, year: int):
Но у меня теперь опять два вопроса. Он эту новую дату считает новым объектом класса Date (как d3) да? Это новый self и именно поэтому он идет в инит?
Вопрос 2. У меня сваливается всё с ругательствами, что нет атрибута _day . Но вот это-то def is_valid_date(cls, day: int, month: int, year: int): принимает просто day,
то есть кому и по какой причине понадобился _day для валидации 31 февраля, от меня ускользает.

**Comment by uliss4ever** [comment]
И вот ещё непонятно и в этих доках, и в декораторе, почему именно такой порядок. Сперва покажи, какой у нас атрибут(атрибут же?), геттером. А потом установи, сеттером. Логичнее же сперва что-то создать(установить новое), а потом это показать.
Самое непонятное на данный момент сейчас. Нам говорили, что в ините всё устанавливается один раз. Попадаем ли мы в инит при установке нового значения через сеттер? И почему? Это инициализация нового значения, которое мы пытаемся установить? (миллион извинений, если я всё назвала не своими именами, плачут коза и слон...)

**Comment by Lex-2008** [comment]
> У нас есть дата 31/01/2021. Она задана в ините (да?) и 31 - это _day, 01 - _month, 2021 -_year, только устанавливали в обратном порядке.
> И мы ему сказали, проверь, пожалуйста, 31/02/2021 будет ли валидной датой? Для этого нужен опять сеттер. Который идет и вызывает вот это def is_valid_date(cls, day: int, month: int, year: int):

Ты всё объясняешь правильно, но ты объясняешь, как его _надо_ написать, а не как оно сейчас написано :)

Я хочу тебе сначала объяснить, почему оно сейчас не работает. Или ты уже готова написать, как оно должно работать?

Возьми листок бумаги, и нарисуй на нем три квадратика с названиями: "_day", "_month", "_year". В квадратиках пока пусто - то есть, ничего.

Теперь представь себя интерпретатором Питона, который входит в функцию init, и несёт в руках список из трех чисел (args). Он видит строчку `self.year = int(args[2])` и понимает: надо пойти в сеттер year и принести туда ~второе~ третье число из тех, что у него в руках.

На листке бумаги ничего не изменилось - по-прежнему три пустых квадратика.

Теперь интерпретатор Питона входит в функцию-сеттер year, неся в руках одно число. Там написано:

>         if not self.is_valid_date(self.day, self.month, value):

Это значит: вычислить `self.day`, вычислить `self.month`, и пойти в функцию `is_valid_date`, неся в руках три числа.

На листке бумаги ничего не изменилось - по-прежнему три пустых квадратика.

`self.day` - это свойство. Чтобы его вычислить, надо сходить в функцию-геттер. Она у нас простая:

````
    def day(self):
        return self._day
````

Расскажи: что интерпретатор Питона будет делать, когда войдёт в неё? Что написано на листке бумаги и что требует сделать функция?

**Comment by Lex-2008** [comment]
> Попадаем ли мы в инит при установке нового значения через сеттер? И почему?

Нет, в инит мы попадаем только тогда, когда запускаем Date с большой буквы.

а вот инит у нас сейчас использует (запускает) сеттеры, и в этом проблема ;)

**Comment by uliss4ever** [comment]
>Что в ней записано? Что надо вычислить "int(args[2])", и записать результат в "self.year". "self" - это у нас объект класса Date (мы сейчас init разбираем, помнишь?), и year - это его свойство. Свойство - значит, когда мы пытаемся туда что-то записать, интерпретатор вызывает сеттер. Сеттер у нас начинается такими словами:
    @year.setter 
    def year(self, value: int): 

Я сейчас остальное дочитаю, пока вот. Это мне очень-очень надо знать. Вот когда интерпретатор видит self.year , он идет в сеттер и вот тут @year.setter именно тот же самый year ? (я просто из-за миллиона одинаковых названий путаюсь, но многое уже поняла. Например, что day это ни разу не day() (обычно я и так это знаю, но не в этих условиях). 
Но вот тут год в селфе и год перед точкой в сеттере напрямую связаны? (мой ответ да, потому что интерпретатор как иначе поймет, какой из трех сеттеров сейчас ему нужен.

**Comment by uliss4ever** [comment]
> self.day - это свойство. Чтобы его вычислить, надо сходить в функцию-геттер. Она у нас простая:

    def day(self):
        return self._day

Хммм. То есть пока он занимается годом, он day через сеттер не прогоняет, а берёт его напрямую из геттера? Что дали, то и берет, непроверенное? Но так не может быть, потому что геттер возвращает _day, а про это _day у нас сказано только в нижележащем сеттере, то есть он идет в сеттер, чтобы проверить день, чтобы вычислить его для того, чтобы проверить валидность даты для отработки year. Но в сеттере дня у нас снова  is_valid_date(value, self.month, self.year) но уже с value  для дня. И так до бесконечности...? Полагаю что нет, поскольку всё работает и не сваливается.

**Comment by Lex-2008** [comment]
> всё работает и не сваливается

Эй, подожди, ты исправила, что ли? Как я тебе буду объяснять, почему оно **не** работает, если оно уже работает?! :D

> пока он занимается годом, он day через сеттер не прогоняет, а берёт его напрямую из геттера?

Да. `day` без подчеркивания - это свойство. Его можно либо брать (читать), либо устанавливать (писать). Если брать-читать - то идут в геттер, если устанавливать-писать, то сеттер. Каждый раз, когда встречаешь "day" без подчеркивания - проверь, его здесь берут или устанавливают.

> Что дали, то и берет, непроверенное?

Да, именно так. Вот что у тебя на листочке написано в квадратике "_day", то и берет. А что у тебя там написано? Ничего? Вот он и валится с криком "_day не установлено".

Если мы с тобой ещё на одной странице, то это я всё объясняю ошибку на том скриншоте, который ты мне сегодня утром по почте прислала. Ну как утром... Файл называется "Снимок экрана 2021-02-28 в 13.47.56.png". Там написано, как интерпретатор питона шел по программе: сначала он выполнял `main()`, потом дошел до строчки `d1 = Date(30, 12, 2021)` и вошел в init. Там дошел до строчки `self.year = int(args[2])` и перешел в функцию `year`, в которой дошел до строки `if not self.is_valid_date(self.day, self.month, value):`, и вошел в функцию `day`, где споткнулся на строчке `return self._day`. И умер с криком "у этого объекта класса Date нету (не установлено) _day".

> Но так не может быть, потому что геттер возвращает _day, а про это _day у нас сказано только в нижележащем сеттере, то есть он идет в сеттер, чтобы проверить день

Нет, Питон не настолько умный :) Как ты помнишь, когда интерпретатор входил в сеттер для года, он нес в руках только одно число - номер года. Номер дня остался в init'е. Там, где мы сейчас - ни в сеттере для года, ни в геттере для дня - номера дня уже нет. Поэтому проверить день нельзя, т.к. его никто не знает.

**Comment by uliss4ever** [comment]
Ненене, я ничего не исправляла, всё сваливается!
Я вчера (если ничего не путаю) поменяла сеттеры для месяца и года. И инит свалился. Лёш, у меня такие чудеса в пайчарме творятся, что вполне может быть не та версия, надо как-то проверить. Я тебе ещё не всё рассказала про этот травис: не работало, потому что yml файл был не в корне, я его перенесла, сделала коммит, заработало (ну отчасти), я внесла ещё поправки, сделала коммит - а в нем yml опять не в корне! Хотя я точно его обратно не переносила (зачем?). Короче помимо моего тупизма ещё какие-то внешние силы похоже вмешиваются))
Где и что в итоге работает?
Оййй, короче. Это вообще про то, что я описала потенциально бесконечный цикл, но так же не происходит, в теории. Всё как надо работает, вот про это я))) Извини, выражаюсь запутано...

**Comment by uliss4ever** [comment]
> Что дали, то и берет, непроверенное?
Да, именно так. Вот что у тебя на листочке написано в квадратике "_day", то и берет. А что у тебя там написано? 

Ничего? Вот он и валится с криком "_day не установлено".
Вот оно! Наконец-то! А то, если честно, у меня у самой такая же реакция - мы же ничего вообще про этот _day ни сном ни духом... 
осталось придумать как сделать чтобы это исправить. И я попытаюсь)

**Comment by Lex-2008** [comment]
Ну попытайся! Подсказка: когда не будет получаться сделать это _с помощью чего-то_, то попробуй сделать то же самое, но без этого _чего-то_! (надеюсь, эта подсказка поможет, а не запутает...)

**Comment by uliss4ever** [comment]
В общем, я понимаю, что надо как-то сделать, чтобы он в проперти за днём не пошёл. Или где-то ему сразу _day_value написать. В общем, пытаюсь додуматься)

**Comment by Lex-2008** [comment]
Да, мне кажется, это ты в правильном направлении думаешь :)

**Comment by uliss4ever** [comment]
Попробовала ещё в инит вот такую проверку сделать check_values = self.is_valid_date(args[0], args[1], args[2]), и в сеттерах её тоже произвести. Но всё как валилось так и валится. Правда перемещалось на 37 строку.
Ещё я думала, может тогда порядок сделать соответственно порядку в ините. Но по-честному пока все идеи кончились. Пока погляжу запись про список, там ещё Енот сказал со слабой ссылкой разобраться, чтобы мне завтра пеньком не сидеть) И драйвера с паттернами ещё. А потом подумаю опять, а может мозг там сам как-нить, без меня разберётся))))

**Comment by Lex-2008** [comment]
> в инит вот такую проверку сделать check_values = self.is_valid_date(…)

Да, это правильно. И что ты потом с check_values будешь делаешь?

Дальше идея должна быть такая: сеттеры делают две вещи: 1) проверяют правильность даты (is_valid_date); 2) если дата правильная, то сохраняют новое значение в переменную _day_value / _month_value / _year_value.

А раз мы уже проверку сделали (и только в том случае, если она показала, что дата верна!), то зачем нам использовать сеттеры? Давай тогда просто запишем день, месяц, год, которые нам передали (args[...]), прямо в переменные _day_value, _month_value, _year_value.

слабые ссылки? драйверы? паттерны? Это всё что-то новое для меня! :)

**Comment by uliss4ever** [comment]
Драйверы и паттерны - это с последней практики, это я сама попробую написать для односвязного пока) Потом надо прикрутить к двусвязному, Енот сказал, это копипастом. А вот слабая ссылка - это надо реализовать в двусвязном списке. Он объяснения давал на другом примере (есть в записи, но сама найду и переслушаю и думаю, мы вставим). Там основное - надо реализовать или переопределить все методы для двусвязного списка, вот это главное. Ну и среди них почему-то понадобится использовать эту ссыль, чтобы не зациклиться, как я поняла) Это в послденей части записи от 19 числа, начинается историей про garbage collection. Щас, короче, послушаю сама...
> А раз мы уже проверку сделали (и только в том случае, если она показала, что дата верна!), то зачем нам использовать сеттеры? Давай тогда просто запишем день, месяц, год, которые нам передали (args[...]), прямо в переменные _day_value, _month_value, _year_value.

Вот я мучаюсь именно поэтому. Мне надо куда-то эти новые значения записать. А он сразу красным всё подчёркивает, не знаю никаких переменных, моя твоя непонимай...
Это первое. А второе - а как же,  если напрямую, и тогда зачем нам сеттеры вообще? Я только-только с ними разобралась)

**Comment by Lex-2008** [comment]
> Мне надо куда-то эти новые значения записать.
Ну а как сеттеры это делают? Вот и инит должен так же делать!
В каждом сеттере есть один оператор присвоения `=`. Что стоит слева от него?

> зачем нам сеттеры вообще? Я только-только с ними разобралась)

А, это просто другой интерфейс, другой способ установки даты. Можно сразу создать дату с нужными днем/месяцем/годом: `d1 = Date(1, 3, 2021)`, а можно сначала создать дату "1 января 0 года", а потом изменять её - устанавливать день, месяц, год:
```
d1 = Date(1, 1 ,1)
d1.month = 3
d1.year = 2021
```

**Comment by uliss4ever** [comment]
Я там новый коммит сделала. Теперь свалилась с out of range. Разнообразие))) У нас было это, щас пойду по следам подумаю. 
Но меня беспокоит, сеттеров вообще не будет? Как-то жалко их)

**Comment by Lex-2008** [comment]
Нет, сеттеры будут, они нам нужны для второго способа установки даты. Если программист напишет:
```
d1 = Date(1, 1 ,1)
d1.month = 3
d1.year = 2021
```

то в первой строчке вызовется инит, а во второй и третьей - сеттеры.

**Comment by uliss4ever** [comment]
Теперь я свалилась вот так)) Строку он с числом сравнить не сможет. А раньше прекрасно мог. Month - это же 
![Снимок экрана 2021-03-01 в 18 01 45](https://user-images.githubusercontent.com/32711430/109515720-71fe6280-7ab8-11eb-9165-c6e3ef4b09dc.png)
Я наставила интов, сравнил. И свалились иначе. Ой, всё(
![Снимок экрана 2021-03-01 в 18 05 19](https://user-images.githubusercontent.com/32711430/109516043-d02b4580-7ab8-11eb-8229-77d6c3c6d4bb.png)

**Comment by Lex-2008** [comment]
ну если ты тут пишешь int(…) вокруг каждого month и day, то почему ты решила их не писать вокруг month и year передаваемых в get_max_day? Но вообще это неправильно.

> Строку он с числом сравнить не сможет. А раньше прекрасно мог

Нет-нет-нет, раньше ему init давал число! А теперь стал давать строку. Так что давай лучше исправим init!

**Comment by uliss4ever** [comment]
> ну если ты тут пишешь int(…) вокруг каждого month и day, то почему ты решила их не писать вокруг month и year передаваемых в get_max_day? Но вообще это неправильно.

Знаешь, что хорошего точно? Я когда творю фигню, то как будто уже знаю, что не надо бы так. 
Пришла Гатамелата и вызвала мировые часы, экранное время и ещё пару вещей. Щас код исправит...

**Comment by uliss4ever** [comment]
Лекция по веб. Слова вроде знакомые. Сервер, например. И совсем незнакомые. Статические данные.
![Снимок экрана 2021-03-01 в 20 10 15](https://user-images.githubusercontent.com/32711430/109532631-5c923400-7aca-11eb-8e7d-75476736fbf7.png)
